light_control.h

```diff
@@ -0,0 +1,110 @@
+#include <Adafruit_NeoPixel.h>
+#include <time.h>
+
+#define SENSOR_PIN  6
+#define PIN 8   // light control pin
+#define PIN_NUM 2 //count of leds
+
+#define WAIT_FOR_SOMEONE 0
+#define INCREASE_BRIGHTNESS 1
+#define LIGHT_ON_WAIT 2
+#define DECREASE_BRIGHTNESS 3
+
+const int statesCount = 4;
+
+/*
+ current state of system:
+ 0 -- waiting to turn light
+ 1 -- turning on light
+ 2 -- waiting to turn off light
+ 3 -- smoothly turn off light and checking sensor and keep illuminate if something detected
+ */
+int state = WAIT_FOR_SOMEONE;
+
+
+class Light_control{
+private:
+    /*
+     time shifts between iterations in each state in mills
+     in state 0 we just waiting for sensor activation, that's why states_timers[0] = NULL
+     in state 1 we have 170 brightness increesing with delay 100 mills
+     in state 2 we one iteration with waiting long time
+     in state 3 we have 170 brightness decreesing and checking sensor and keep illuminate if something detected
+     */
+    const unsigned long states_timers[statesCount] = {NULL, 100, 240000, 100};
+    unsigned long states_starts[statesCount] = {NULL, 0, 0, 0};
+    Adafruit_NeoPixel strip;
+    const int brightness_shift = 75;
+    int brightness;
+    unsigned long current_time;
+    
+    int brightness_change(bool upDown){
+        int result = upDown ? INCREASE_BRIGHTNESS : DECREASE_BRIGHTNESS;
+        int change_val = (current_time - states_starts[result])/states_timers[result];
+        if (!(change_val >= 1)){
+            return result;
+        }
+        brightness += upDown ? change_val : -1 * change_val;
+        states_starts[result] = current_time;
+        int new_val = brightness + brightness_shift;
+        strip.setPixelColor(0, strip.Color(new_val, new_val, 0));
+        strip.show();
+        if (upDown){
```

Comment:
I suggest to change the code a bit:

```c++
const int maximum_brightness = 170;

bool reached_maximum_brightness(int state, int brightness) {
    return (state == INCREASE_BRIGHTNESS) and (brightness >= maximum_brightness)       
}

bool reached_minimum_brightness(int state, int brightness) {
    return (state == DECREASE_BRIGHTNESS) and (brightness <= 0)
}
```

then here it would be more compact

```c++
if reached_maximum_brightness(state, brightness){
    return LIGHT_ON_WAIT;
}
if reached_minimum_brightness(state, brightness){
    return WAIT_FOR_SOMEONE;
}
```

Reply:
You right, it made code more clear